DicomExtract

catgonzftw1 opened the pull request:
I attempted to add code that extracts from a DICOM file.

catgonzftw1 committed:
attempted to add DICOM Extraction

catgonzftw1 commented:
```
def extract_dicom_metadata(found_files: list) -> dict:
    dicom_metadata = {}  # Initialize the metadata dictionary
    
    for file_name in found_files:
        if file_name.lower().endswith('.dcm'):
            ds = pydicom.dcmread(file_name)
            for elem in ds.iterall():
                if elem.tag.is_private:
                    continue  # Skip private tags
                tag_str = str(elem.tag)
                value = elem.value
                if isinstance(value, pydicom.valuerep.DSfloat):
                    value = float(value)
                dicom_metadata[tag_str] = value

    return dicom_metadata

dicom_metadata = extract_dicom_metadata(found_files)

```

I think the above is the pertinent code. It used the list of found files retrieved by scan_directory()

```
def scan_directory(directory: str, file_patterns: list[str]) -> list:
    file_list = []
    for pattern in file_patterns:
        for file in Path(directory).rglob(pattern):
            if file.is_file():
                file_list.append(file.as_posix())
    return file_list

directory = r"C:\Cat\10-19 CRC_1382\13 sarah_preclinical_data"  # Use 'r' prefix to treat backslashes as literal characters
file_patterns = ["*.dcm", "*.mct", "*.parameters"]  # Update with your desired patterns
found_files = scan_directory(directory, file_patterns)
print(found_files)
```

catgonzftw1 committed:
changed file name to match class

catgonzftw1 committed:
adding function for mCT metadata extraction

catgonzftw1 committed:
Add parameter extraction code

BenediktHeinrichs commented on MetadataExtractor/Extractors/Data/IDataExtract.py:
```diff
@@ -3,4 +3,4 @@
 class IDataExtract(IExtract):
     def registerMimeTypes(self):
-        self.mimeTypes["concrete"] = ["text/csv", "text/xml", "application/json"]
+        self.mimeTypes["concrete"] = ["text/csv", "text/xml", "application/json", "dcm"]
```
Please remove that from the default, I don't think this is appropriate as a default value (and won't be necessary for your case).

BenediktHeinrichs commented on MetadataExtractor/Extractors/Data/ParameterExtract.py:
```diff
@@ -0,0 +1,54 @@
+log = logging.getLogger(__name__)
+
+
+class mCTExtract(IDataExtract):
```
This seems to be obsoleted by `DicomExtract`, so please remove that file.

BenediktHeinrichs commented on MetadataExtractor/Extractors/Data/mCTExtract.py:
```diff
@@ -0,0 +1,56 @@
+log = logging.getLogger(__name__)
+
+
+class mCTExtract(IDataExtract):
```
This seems to be obsoleted by `DicomExtract`, so please remove that file.

BenediktHeinrichs commented on MetadataExtractor/Extractors/Data/DicomExtract.py:
```diff
@@ -0,0 +1,57 @@
+        values = []
+        identifier = fileInfo["identifier"]
+
+        for attribute in attributes.keys():
```
you never define any attribute, you should make use of dicom_metadata here instead of attributes

BenediktHeinrichs commented on MetadataExtractor/Extractors/Data/DicomExtract.py:
```diff
@@ -0,0 +1,57 @@
+from .IDataExtract import IDataExtract
+from MetadataExtractor.Util import metadataCreation, metadataFormatter
+import pydicom
```
please add pydicom to the [requirements.txt](https://github.com/BenediktHeinrichs/metadataextractor/blob/main/requirements.txt)

BenediktHeinrichs replied:
This has not been resolved

BenediktHeinrichs commented on MetadataExtractor/Extractors/Data/DicomExtract.py:
```diff
@@ -0,0 +1,57 @@
+log = logging.getLogger(__name__)
+
+
+class DicomExtract(IDataExtract):
```
Make sure to add `DicomExtract` to the following list to ensure it properly working: https://github.com/BenediktHeinrichs/metadataextractor/blob/main/defaultConfigs.py#L73

BenediktHeinrichs replied:
This has not been resolved

BenediktHeinrichs reviewed (changes requested):
Hey @catgonzftw1, I reviewed your changes. Still needs a bit of work, feel free to contact me if something is unclear.

catgonzftw1 committed:
Delete MetadataExtractor/Extractors/Data/ParameterExtract.py

removed superfluous file

catgonzftw1 committed:
Delete MetadataExtractor/Extractors/Data/mCTExtract.py

removed superfluous file

catgonzftw1 reviewed (commented):
removed the mCT and parameter extraction files

catgonzftw1 committed:
Update IDataExtract.py

Removed , "dcm" from default

catgonzftw1 reviewed (commented):
Removed the default value as requested

catgonzftw1 committed:
Updated DicomExtract.py to use dicom_metadata

Removed the undefined attribute parameter.

catgonzftw1 reviewed (commented):
removed undefined attribute parameter

catgonzftw1 commented:
@BenediktHeinrichs I made changes as you noted. Please let me know if it still needs something changed. I can give it another shot :)

BenediktHeinrichs reviewed (changes requested):
@catgonzftw1 I think you missed some comments. Let's talk about this tomorrow.

catgonzftw1 commented:
@BenediktHeinrichs I was able to edit those files and create pull requests for the unfinished changes. Please let me know if it looks correct :)

catgonzftw1 committed:
adding pydicom, adding DicomExtract to config
    new file:   Examples/2022-03-07_18h21_M3_Cy7.mCT
	new file:   Examples/2022-03-07_18h21_M3_Cy7.parameters
	new file:   Examples/EnIm1.dcm
	modified:   MetadataExtractor/Extractors/Data/DicomExtract.py
	modified:   MetadataExtractor/Extractors/Data/IDataExtract.py
	deleted:    MetadataExtractor/Extractors/Data/ParameterExtract.py
	deleted:    MetadataExtractor/Extractors/Data/mCTExtract.py
	modified:   defaultConfigs.py
	modified:   requirements.txt

catgonzftw1 committed:
adding updates

catgonzftw1 reviewed (commented):
Removed files and made the requested changes into one pull request

catgonzftw1 commented:
@BenediktHeinrichs I moved all the changes into this pull request but when I ran the file it throws an error in the first line

`from .IDataExtract import IDataExtract`

The codespace suggested I make a virtual environment which I did but still get the import error:

`ImportError: attempted relative import with no known parent package`

BenediktHeinrichs commented on MetadataExtractor/Extractors/Data/DicomExtract.py:
```diff
@@ -0,0 +1,59 @@
+        return (text, metadata)
+
+    def registerMimeTypes(self):
+        self.mimeTypes["concrete"] = ["application/dcm"]
```
Add "application/dicom" as the mimetype

BenediktHeinrichs commented on MetadataExtractor/Extractors/Data/DicomExtract.py:
```diff
@@ -0,0 +1,59 @@
+        log.info('Extracting metadata for Dicom file "' + fileInfo["file"] + '".')
+        text = ""
+
+        dicom_metadata = pydicom.dcmread(file)
```
Change `file` to `fileInfo["file"]`

BenediktHeinrichs replied:
On parsing, I get the following error, please check why this is the case & resolve it: 

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/pydicom/dataset.py", line 2283, in __setitem__
    key = Tag(key)
  File "/usr/local/lib/python3.10/site-packages/pydicom/tag.py", line 122, in Tag
    raise ValueError(
ValueError: Unable to create an element tag from '(0008, 0008)': unknown DICOM element keyword or an invalid int
```

BenediktHeinrichs reviewed (changes requested):
@catgonzftw1 I provided some additional feedback. Please make sure to run & test the code before requesting another review. E.g., you can do this by executing python pipeline_runner.py & checking the output folder afterwards.

The file also isn't meant to run individually. If you still want to do that, check Image/ObjectExtract and make sure to run the file on the correct file level (otherwise the imports are not found). (cd to the correct directory and then execute the python file)

catgonzftw1 committed:
dicom extractor appears to work now

catgonzftw1 commented:
@BenediktHeinrichs I made some changes and tested it. I think it's working now.

BenediktHeinrichs commented:
@catgonzftw1 extraction seems to work now. Is this what you envisoned?

![image](https://github.com/BenediktHeinrichs/metadataextractor/assets/49311227/d4e4b914-7eab-482c-a52a-5eba3711005f)

If yes, just briefly notify me and I'll merge this.
